### ISSUE TITLE
Bump `vmm-sys-util` to v0.12.0 or newer

### ISSUE THREAD

**Issue by likebreath** [cloud-hypervisor/cloud-hypervisor#6072]
_Originally posted by @likebreath in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6068#issuecomment-1874685576_:

The `vmm-sys-util` is indeed being used by a lot other dependencies. We will need to wait for these dependencies to be updated with the new version of `vmm-sys-util`. Let's hold on the version bump together with other rust-vmm version bump. 

Note that the new release of `vmm-sys-util` v0.12.0+ includes patches for https://github.com/rust-vmm/vmm-sys-util/security/advisories/GHSA-875g-mfp6-g7f9.

**Comment by jinankjain** [comment]
This one is bit complicated, we would also need to update the internal fork of kvm-bindings to the latest version.

**Comment by rbradford** [comment]
Here is the dependency tree:

```rockhopper:~/src/cloud-hypervisor (main)$ cargo tree -i vmm-sys-util@0.11.1 | grep -v "/home/rob"
vmm-sys-util v0.11.1
├── kvm-bindings v0.6.0 (https://github.com/cloud-hypervisor/kvm-bindings?branch=ch-v0.6.0-tdx#7d9ffb47)
│   ├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04)
│   │   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
│   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04) (*)
│   │   [dev-dependencies]
│   [dev-dependencies]
├── versionize v0.1.10
├── vfio-bindings v0.4.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
│   ├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
│   └── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997)
├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
├── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997) (*)
├── vhost v0.9.0
├── virtio-queue v0.10.0
```

vfio needs this PR merged https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6072

And we need to update kvm-bindings fork: I suggest we stop trying to use autogenerated TDX constants and just pull those into the arch crate & then just port the serialisation support atop the latest released kvm-bindings. @likebreath are you able to handle this?

**Comment by likebreath** [comment]
> Here is the dependency tree:
> 
> ```
> vmm-sys-util v0.11.1
> ├── kvm-bindings v0.6.0 (https://github.com/cloud-hypervisor/kvm-bindings?branch=ch-v0.6.0-tdx#7d9ffb47)
> │   ├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04)
> │   │   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
> │   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> ├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04) (*)
> │   │   [dev-dependencies]
> │   [dev-dependencies]
> ├── versionize v0.1.10
> ├── vfio-bindings v0.4.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
> │   ├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> │   └── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997)
> ├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> ├── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997) (*)
> ├── vhost v0.9.0
> ├── virtio-queue v0.10.0
> ```
> 

Likely the new versions of these crates are using new vm-memory, so basically all rust-vmm crates are going to be updated together.

> vfio needs this PR merged #6072
> 

Sorry, what do you mean by this?

> And we need to update kvm-bindings fork: I suggest we stop trying to use autogenerated TDX constants in our fork and just pull those into the arch crate (i.e. copying the ones currently needed) & then just port the serialisation support atop the latest released kvm-bindings. 

We need more than just constants from the `kvm-binding`s fork, such as new KVM data structures and KVM APIs. It would be pretty hard to pull all these required bits into the arch crate.

I can see that the latest `kvm-ioctl` crate won't work with our current `kvm-binding`s fork (as they have different kernel base), but I think it will still be easier to update our `kvm-bindings` fork and make it work. I will have to try it out.

> @likebreath are you able to handle this?

Yes, I will have time for this later next week.

**Comment by rbradford** [comment]
> > Here is the dependency tree:
> > ```
> > vmm-sys-util v0.11.1
> > ├── kvm-bindings v0.6.0 (https://github.com/cloud-hypervisor/kvm-bindings?branch=ch-v0.6.0-tdx#7d9ffb47)
> > │   ├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04)
> > │   │   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
> > │   └── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> > ├── kvm-ioctls v0.13.0 (https://github.com/rust-vmm/kvm-ioctls?branch=main#23a3bb04) (*)
> > │   │   [dev-dependencies]
> > │   [dev-dependencies]
> > ├── versionize v0.1.10
> > ├── vfio-bindings v0.4.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa)
> > │   ├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> > │   └── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997)
> > ├── vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#59c604fa) (*)
> > ├── vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#6c72e997) (*)
> > ├── vhost v0.9.0
> > ├── virtio-queue v0.10.0
> > ```
> 
> Likely the new versions of these crates are using new vm-memory, so basically all rust-vmm crates are going to be updated together.
> 
> > vfio needs this PR merged #6072
> 
> Sorry, what do you mean by this?
> 

I was trying to succinctly (too succinctly :-) ) say that all the rest of the crates have been updated and released with the new dependency added. 

> > And we need to update kvm-bindings fork: I suggest we stop trying to use autogenerated TDX constants in our fork and just pull those into the arch crate (i.e. copying the ones currently needed) & then just port the serialisation support atop the latest released kvm-bindings.
> 
> We need more than just constants from the `kvm-binding`s fork, such as new KVM data structures and KVM APIs. It would be pretty hard to pull all these required bits into the arch crate.
> 
> I can see that the latest `kvm-ioctl` crate won't work with our current `kvm-binding`s fork (as they have different kernel base), but I think it will still be easier to update our `kvm-bindings` fork and make it work. I will have to try it out.
> 

I was going with the approach of what I thought was the least work - now and later. Having a fork of kvm-bindings for serialisation is annoying, having to generate a completely different set of headers seems tedious. If you think it's less work to generate new headers and patch then that's fine. But we really want to move to modern kvm-ioctls.

> > @likebreath are you able to handle this?
> 
> Yes, I will have time for this later next week.
